restore/data.go

```diff
@@ -220,6 +227,27 @@ func restoreDataFromTimestamp(fpInfo filepath.FilePathInfo, dataEntries []toc.Co
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel() // Make sure it's called to release resources even if no errors
 
+	// Launch a checker that polls if the restore helper has ended with an error.
+	// It is our 'Ultima ratio regum' - in case restore helper couldn't read a file with the oid list, it is not aware
+	// about the pipes, pre-created by the gprestore, and it can't close them.  So we cancel all pending COPY commands
+	// from here after giving a chance to the restore helper to close pipes on its own.
+	if backupConfig.SingleDataFile || resizeCluster {
+		go func() {
+			for {
+				time.Sleep(5 * time.Second)
+				remoteOutput := globalCluster.GenerateAndExecuteCommand("Checking gpbackup_helper agent failure", cluster.ON_SEGMENTS, func(contentID int) string {
+					helperErrorFileName := fmt.Sprintf("%s_error", fpInfo.GetSegmentPipeFilePath(contentID))
+					return fmt.Sprintf("! ls %s", helperErrorFileName)
+				})
+				if remoteOutput.NumErrors != 0 {
+					// the delay below is to give the restore helper a chance to close all pipes, if it can...
+					time.Sleep(5 * time.Second)
+					cancel()
+				}
+			}
+		}()
```

Comment:
> This logic works fine, but it will run for the entire duration of the program, and I think it's a little overkill :c. 
> Could this be simplified/made lighther? Like using some kind of gpbackup_helper barrier, for example, wait until gpbackup_helper creates a ready file on all segments, and avoid the asynchronous "supervisor loop"?

I think it is better to keep the code more simple than adding additional "sync" files to helper just to make one goroutine (running once in several seconds) finish earlier.


> Is it justified and catches most of the possible helper failures?

Yes, it should catch helper failures, as the helper creates "_error" file in case of any error and if it was terminated by a signal.